### ISSUE TITLE
Pretrained checkpoint not loading with FSDP2 multi-GPU (AutoModel + Qwen3 base; custom combined-projection model)

### ISSUE THREAD

**Issue by zjr2000** [NVIDIA-NeMo/Automodel#1148]
**Describe the bug**

When running SFT on Qwen3 base using AutoModel + FSDP2 across multiple GPUs, the loss is extremely high, suggesting the pretrained weights are not being loaded (model appears randomly initialized).

Similarly, with a custom model that uses combined projections (merged QKV and MLP projections), FSDP2 multi-GPU fails to load the pretrained checkpoint, but the same checkpoint loads correctly using the naive Hugging Face (non-FSDP) implementation.

**Steps/Code to reproduce bug**
The used data is a small subset of smoltalk.
```
step_scheduler:
  global_batch_size: 2
  local_batch_size: 1
  ckpt_every_steps: 200
  val_every_steps: 100  # will run every x number of gradient steps
  num_epochs: 1

dist_env:
  backend: nccl
  timeout_minutes: 1

rng:
  _target_: nemo_automodel.components.training.rng.StatefulRNG
  seed: 1111
  ranked: true

model:
  _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
  pretrained_model_name_or_path: checkpoints/Qwen3-1.7B-Base
  _attn_implementation: sdpa

peft:
  _target_: nemo_automodel.components._peft.lora.PeftConfig
  match_all_linear: True
  dim: 32
  alpha: 32
  use_triton: False

checkpoint:
  enabled: false
  checkpoint_dir: output/qwen3_1_7b_peft/
  model_save_format: safetensors # torch_save or safetensors
  save_consolidated: true # saves the model in a consolidated safetensors format. Requires model_save_format to be safetensors.

distributed:
  _target_: nemo_automodel.components.distributed.fsdp2.FSDP2Manager
  dp_size: none
  tp_size: 1
  cp_size: 1
  sequence_parallel: false


loss_fn:
  _target_: nemo_automodel.components.loss.linear_ce.FusedLinearCrossEntropy

dataset:
  _target_: nemo_automodel.components.datasets.llm.chat_dataset.ChatDataset
  path_or_dataset_id: data/sft_debug_200.jsonl
  split: "train"
  shift_labels: false
  seq_length: 4096
  truncation: true
  tokenizer:
    _target_: transformers.AutoTokenizer.from_pretrained
    pretrained_model_name_or_path: checkpoints/Qwen3-1.7B-Base

packed_sequence:
  # Set packed_sequence_size > 0 to run with packed sequences
  packed_sequence_size: 0

dataloader:
  _target_: torchdata.stateful_dataloader.StatefulDataLoader
  collate_fn: nemo_automodel.components.datasets.utils.default_collater
  shuffle: true


optimizer:
  _target_: torch.optim.Adam
  betas: [0.9, 0.999]
  eps: 1e-8
  lr: 1.0e-5
  weight_decay: 0
  # min_lr: 1.0e-5
```
Loss
```
2026-02-02 17:46:14 | INFO | root | step 0 | epoch 0 | loss 14.4183 | grad_norm 25.5936 | lr 1.00e-05 | mem 14.02 GiB | tps 496.03(248.02/gpu) | num_label_tokens 5628
2026-02-02 17:46:22 | INFO | root | step 1 | epoch 0 | loss 14.9602 | grad_norm 24.4426 | lr 1.00e-05 | mem 11.98 GiB | tps 849.40(424.70/gpu) | num_label_tokens 5784
2026-02-02 17:46:30 | INFO | root | step 2 | epoch 0 | loss 12.7187 | grad_norm 19.7951 | lr 1.00e-05 | mem 10.31 GiB | tps 557.32(278.66/gpu) | num_label_tokens 3805
2026-02-02 17:46:40 | INFO | root | step 3 | epoch 0 | loss 13.1190 | grad_norm 23.0947 | lr 1.00e-05 | mem 9.82 GiB | tps 604.38(302.19/gpu) | num_label_tokens 5040
2026-02-02 17:46:51 | INFO | root | step 4 | epoch 0 | loss 13.8348 | grad_norm 22.2743 | lr 1.00e-05 | mem 8.05 GiB | tps 417.45(208.72/gpu) | num_label_tokens 2833
2026-02-02 17:47:01 | INFO | root | step 5 | epoch 0 | loss 14.4640 | grad_norm 25.4368 | lr 1.00e-05 | mem 14.09 GiB | tps 640.34(320.17/gpu) | num_label_tokens 5130
```

**Comment by ZhiyuLi-Nvidia** [comment]
Hi, I took a try and seemed not be able to reproduce the above issue in current `main` branch:

I was using the following recipe as I have no access to your dataset.
```yaml
step_scheduler:
  global_batch_size: 2
  local_batch_size: 1
  ckpt_every_steps: 200
  val_every_steps: 100  # will run every x number of gradient steps
  num_epochs: 1
  max_steps: 10

dist_env:
  backend: nccl
  timeout_minutes: 1

rng:
  _target_: nemo_automodel.components.training.rng.StatefulRNG
  seed: 1111
  ranked: true

model:
  _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
  pretrained_model_name_or_path: Qwen/Qwen3-1.7B-Base
  force_hf: true  # turn on/off naive hf implementation
  _attn_implementation: sdpa
  output_hidden_states: true

peft:
  _target_: nemo_automodel.components._peft.lora.PeftConfig
  match_all_linear: True
  dim: 32
  alpha: 32
  use_triton: False

checkpoint:
  enabled: false
  checkpoint_dir: output/qwen_debug/
  model_save_format: safetensors # torch_save or safetensors
  save_consolidated: true # saves the model in a consolidated safetensors format. Requires model_save_format to be safetensors.

distributed:
  _target_: nemo_automodel.components.distributed.fsdp2.FSDP2Manager
  dp_size: none
  tp_size: 1
  cp_size: 1
  sequence_parallel: false


loss_fn:
  _target_: nemo_automodel.components.loss.linear_ce.FusedLinearCrossEntropy

dataset:
  _target_: nemo_automodel.components.datasets.llm.squad.make_squad_dataset
  dataset_name: rajpurkar/squad
  split: train

packed_sequence:
  # Set packed_sequence_size > 0 to run with packed sequences
  packed_sequence_size: 4096

dataloader:
  _target_: torchdata.stateful_dataloader.StatefulDataLoader
  collate_fn: nemo_automodel.components.datasets.utils.default_collater
  shuffle: true


optimizer:
  _target_: torch.optim.Adam
  betas: [0.9, 0.999]
  eps: 1e-8
  lr: 1.0e-5
  weight_decay: 0
  # min_lr: 1.0e-5
```

I turn on/off naive HF implementation with `force_hf: true|false` and both loss starts with ~8.xx.
Do you happen to provide the branch you were using?

**Comment by zjr2000** [comment]
Thanks for the quick response! Setting force_hf works for me with my custom model.

Apologies for the confusion earlier: I’ve since confirmed that the high loss in my example wasn’t caused by the checkpoint-loading/FSDP2 issue. It came from my custom label-shifting implementation. Everything works as expected when I use the standard autoregressive training paradigm on the current main  @ZhiyuLi-Nvidia .

**Comment by qiaochuz-nv** [comment]
## QA RCCA Analysis

**Fix PR:** No dedicated fix. Workaround: `force_hf: true`. Engineer confirmed "not able to reproduce on current main."
**Root Cause:** Custom combined-projection model (Qwen3) pretrained weights not loaded on multi-GPU FSDP2 — model appeared random (high loss).

**Coverage Assessment:**

| Dimension | Bug Config | Nearest Test | Match |
|-----------|-----------|--------------|-------|
| Qwen3 model | Required | qwen3_0p6b_fused_ce_test (2 GPUs), qwen3_4b_convergence_* | Yes |
| FSDP2 multi-GPU | Required | All multi-GPU tests | Yes |
| Custom model path | Not force_hf | Qwen tests don't use force_hf (verified) | Yes |
| Weight loading correctness | Silent | qwen3_4b_convergence tests check loss decrease | Partial |

**Status:** PARTIAL — Qwen3 tests exercise custom model + multi-GPU + FSDP2. Convergence tests verify loss decreases (would catch random-weight initialization). However, this was a community-reported issue that engineer couldn't reproduce on current main — likely fixed by broader refactoring.

**Conclusion:** No new test needed. Qwen convergence tests provide regression coverage for weight loading correctness.
